### ISSUE TITLE
Product show in stock even if the salable quantity is 0

### ISSUE THREAD

**Issue by chichitsai0517** [magento/magento2#31117]
### Preconditions (*)

1. Magento Open Source 2.4.1 with sample data
2.

### Steps to reproduce (*)

1. Make an order of a product and buy everything (example if 5 in stock, buy all 5)
2. Go to your admin and this product QTY should be 5 and Salable QTY should be 0. Which is normal since the order has not been shipped yet. 

### Expected result (*)

1. Front end should show "Out of stock"
2. We should NOT have the "add to cart" Button

### Actual result (*)

1. Front end shows "in stock".
2. Add to cart button is available
3. When we try to cart, we get a message "The requested qty is not available"

NOTE:
See the related issues:
- https://github.com/magento/magento2/issues/31267
- MSI Repo https://github.com/magento/inventory/issues/3102
- https://github.com/magento/inventory/issues/3062
- https://github.com/magento/magento2/issues/25877
- MSI repo https://github.com/magento/inventory/issues/2708
- https://github.com/magento/magento2/issues/27223
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by SchranzKatharina** [comment]
Facing the same issue....

**Comment by engcom-Bravo** [comment]
Hello @chichitsai0517 

Thank you for your report

We confirm that the issue exists on 2.4.1 composer version. It may not appear on 2.4-develop because the MSI modules are missing

Related issues:
- https://github.com/magento/magento2/issues/31267

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-39869` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by sidolov** [comment]
@chichitsai0517 do you have all Magento consumers running?

**Comment by engcom-Kilo** [comment]
Hello, @chichitsai0517 
I'm not able to reproduce this issue on 1.2-develop MSI with consumers running.
![PlaceOrderAllQty](https://user-images.githubusercontent.com/60192090/102785212-40756680-43a6-11eb-8a28-43f185acca79.gif)
Looks like this issue has been fixed in https://github.com/magento/inventory/commit/476e0180abb79b683c6579d52c156d1e8136b0b7

**Comment by magento-engcom-team** [comment]
Hi @chichitsai0517, @engcom-Kilo. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-39869` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.

**Comment by s-kerdel** [comment]
Is there a way to cherry-pick this fix? v.2.4.3 sounds far away and this issue is keeping us from going production for over a month now.

**Comment by Quentin-Sch** [comment]
> Is there a way to cherry-pick this fix? v.2.4.3 sounds far away and this issue is keeping us from going production for over a month now.

Up!

**Comment by prashantvictory** [comment]
@magento-engcom-team could you please advise if this can be shipped with upcoming 2.4.2?

**Comment by thai2301** [comment]
Up!

**Comment by james-noon** [comment]
Wow yes please implement into 2.4.2 or at least a hotfix...this is an incredibly annoying big bug. Looks like from past reports this has been a problem since 2.3.x (from this time last year: https://community.magento.com/t5/Magento-2-x-Technical-Issues/Display-Out-of-Stock-when-Salable-Quantity-is-0/td-p/289077)

To have to wait another 6+ months [for 2.4.3] is pretty ridiculous honestly. As @s-kerdel asked, if we must wait, please let us know if there is a way to backport this for 2.4.2.

**Comment by chequille** [comment]
Need this fix as well.
Guys come on an let us know!!!!

**Comment by vcollina** [comment]
Same problem here. Unbelievable such a simple and basic thing as a product going out of stock to be so painfull. We literally have to monitor the products hourly to update them to "out of stock" manually and despite this effort we still have to answer complains from customers. Its a ton of work literally wasted. What a regret having updated from 1.9 to this.

**Comment by sonbui00** [comment]
@magento-engcom-team Are you here!

**Comment by taipt-0504** [comment]
Pls solve this problem. It impact a lot

**Comment by AlexanderHuyghebaert** [comment]
Any update on a cherry-pick of a temp fix untill update is out?

**Comment by enyceedanny** [comment]
Same issue for me on 2.4.2. I understand that quantity isn't updated until order is shipped/complete - but it should still show as out of stock on the frontend.

**Comment by farrukhk** [comment]
We stopped using MSI because of the same bug when MSI was first introduced. It is very unfortunately to see that in such a long time magento was not able to fix this simple issue. We are Enterprise customer but still there is no solution to this problem.
Now with 2.4.2 we have no choice but to use MSI as if we disable MSI, categories can't be accessed in backend.

**Comment by ioweb-gr** [comment]
I don't get why some issues which impact so many users don't get delivered as quality patches

**Comment by amr-oreaba-mok** [comment]
> Hello, @chichitsai0517
> I'm not able to reproduce this issue on 1.2-develop MSI with consumers running.
> ![PlaceOrderAllQty](https://user-images.githubusercontent.com/60192090/102785212-40756680-43a6-11eb-8a28-43f185acca79.gif)
> Looks like this issue has been fixed in [magento/inventory@476e018](https://github.com/magento/inventory/commit/476e0180abb79b683c6579d52c156d1e8136b0b7)

check this reply, may be it fixes it as a workaround.

**Comment by chequille** [comment]
@amr-oreaba-mok 
I changed inventory code with the the link you gave above ( magento/inventory@476e018).
It looks like it will solve the problem when using
    	`$salable = $this->getSalableQuantityDataBySku->execute($sku);`
which I am using in my stock modul (showing it as a traffic light).
But as you see in the picture the selectable swatches still can be selected.
![grafik](https://user-images.githubusercontent.com/15807794/117196167-7449a980-ade6-11eb-9367-372b289f81c4.png)
Will this be correct as well?
BR
Chequille

**Comment by mysil2** [comment]
@chequille I applied a before plugin to GetAllowProducts function in Configurable class (/vendor/magento/module-configurable-product/Block/Product/View/Type/Configurable.php) that I think will solve your issue. Although this doesn't fix the visibility in categories and out of stock status of the product etc.
```php
<?php


namespace [Some namespace for your plugin];

use Magento\Catalog\Model\Product\Attribute\Source\Status;
use Magento\InventoryCatalog\Model\GetStockIdForCurrentWebsite;
use Magento\InventorySales\Model\GetProductSalableQty;
use Magento\InventorySalesApi\Model\GetStockItemDataInterface;

class Configurable
{
    /**
     * @var GetProductSalableQty
     */
    private $getProductSalableQty;

    /**
     * @var GetStockIdForCurrentWebsite
     */
    private $getStockIdForCurrentWebsite;

    /**
     * @var GetStockItemDataInterface
     */
    private $getStockItemData;

    public function __construct(
        GetProductSalableQty $getProductSalableQty,
        GetStockIdForCurrentWebsite $getStockIdForCurrentWebsite,
        GetStockItemDataInterface $getStockItemData)
    {

        $this->getProductSalableQty = $getProductSalableQty;
        $this->getStockIdForCurrentWebsite = $getStockIdForCurrentWebsite;
        $this->getStockItemData = $getStockItemData;
    }

    // Set getAllowProducts before to prevent returning products which is not salable
    public function beforeGetAllowProducts($subject)
    {
        if (!$subject->hasData('allow_products')) {
            $products = [];
            $allProducts = $subject->getProduct()->getTypeInstance()->getUsedProducts($subject->getProduct(), null);
            /** @var $product \Magento\Catalog\Model\Product */
            foreach ($allProducts as $product) {
                $sku = $product->getSku();
                $stockId = $this->getStockIdForCurrentWebsite->execute();
                $productQtyInStock = $this->getProductSalableQty->execute($sku, $stockId);
                if ((int) $product->getStatus() === Status::STATUS_ENABLED && $productQtyInStock > 0) {
                    $products[] = $product;
                }
            }
            $subject->setData('allow_products', $products);
        }
        return [];
    }

}
```

**Comment by chequille** [comment]
@mysil2 
Hi Thomas, thanks for this. Nice solution, I made the plugin and it is working
![grafik](https://user-images.githubusercontent.com/15807794/124367857-456f7880-dc5b-11eb-80c4-703753c6092d.png)
Only thing I would like to have are the crossed out swatches. Your solution does only show the swatches which are salable.
Any idea to change this?
Best regards,
Chequille

**Comment by chequille** [comment]
@mysil2 
Hi again,
unfortunately, your modul does not work if cache is not cleaned. The method beforeGetAllowedProducts is not called if cache is not cleaned after an order was made.

I am really angry now as I am paying for a VSever since half a year and cannot switch to it because of all the bugs in Mangeto 2.4.2

Does somebody know when there will be a 2.4.3 release?
Does somebody may have a solution for all this pain?

Regards,
Chequille

**Comment by mahmoudelbadrymokalmat** [comment]
As a workaround I used the following plugin to disable MSI and deduct directly from stock when ordering a product 
https://github.com/AmpersandHQ/magento2-disable-stock-reservation/
Because our use case only has a single inventory for the store

**Comment by mysil2** [comment]
@chequille ah, I see - we're doing automatic block cache cleans every 1 or 2 minutes or so because we always had problems getting stock to update as fast as we want - so I didn't catch that.

As of showing the swatches, I guess that's because those are rendered client side and my plugin only sends the salable options to the client. In order to also send configurations that is out of stock, you'll need to have a plugin for the GetJsonConfig function as well (in same class Configurable.php). This is what I do to show the right salable qty available on my product page:
```php
    // Adding Quantitites (product=>qty)
    public function aroundGetJsonConfig(
        \Magento\ConfigurableProduct\Block\Product\View\Type\Configurable $subject, \Closure $proceed)
    {
        $quantities = [];
        $config = $proceed();
        $config = $this->jsonDecoder->decode($config);

        foreach ($subject->getAllowProducts() as $product) {
            $sku = $product->getSku();
            $stockId = $this->getStockIdForCurrentWebsite->execute();
            $productQtyInStock = $this->getProductSalableQty->execute($sku, $stockId);
            $quantities[$product->getId()] = $productQtyInStock;
        }

        $config['quantities'] = $quantities;

        return $this->jsonEncoder->encode($config);
    }
```
I guess you could switch out getAllowProducts() with a function to get all products. And then add desired data to the $config parameter. Further you'll need to override your swatch-renderer.js in order to use the added data as you want. In my custom swatch-renderer.js I get the added data this way: ``var qty = $widget.options.jsonConfig.quantities[this.getProduct()];``

As for 2.4.3, I think it's scheduled for August 10.

**Comment by chequille** [comment]
@mysil2 
Hi, will try this, but where to change the swatch-renderer.js?
I did not find qty in the existing code of swatch-renderer.js

And how do you make this automatic block caceh cleans?

Could you give me some hints on that?

**Comment by mysil2** [comment]
@chequille qty is my addition, so you wouldn't find it in the original swatch-renderer.js which is located in /vendor/magento/module-swatches/view/base/web/js/swatch-renderer.js

In my module I've copied it to /My/Module/view/base/web/js/swatch-renderer.js and added the following in /My/Module/view/base/requirejs-config.js
```js
var config = {
    map: {
        '*': {
            'Magento_Swatches/js/swatch-renderer':'My_Module/js/swatch-renderer'
        }
    }
};
```

**Comment by chequille** [comment]
@mysil2 
Ok, thanks for that info.
To include an own js file is clear, I know that. But the question is what are you doing with your addition var qty?
Would you share your swatch-renderer.js file?
Let me know.
Thanks and BR
Chequille

**Comment by mysil2** [comment]
@chequille 
Just showing available stock when customer choose variations. It's only an example of what can be done, not a solution you're seeking if I understand correctly.

**Comment by chequille** [comment]
Ok, so you do not have a modified swatch-renderer.js where you are using the available stock?
I just do not know where I could use this qty which you mentioned above.
Maybe finally I have to wait some more days till 10th of august with the hope, that 2.4.3 fixed this stock issues.
If not I will be very angry as I am paying for a VServer since 9 months now but could not switch to 2.4.x because of all this bugs.
But, it would not help at the end because nobody cares as I believe sometimes.
BR
Chequille

**Comment by chequille** [comment]
Unfortunately, this issue is not solved for configurable products.
The options are still not crossed if salable quantity is 0.
This is very very very bad.
Sorry guys, what is going on here. Are you not able to fix this? It is a mandatory feature and prevents my shop to go online with magento 2.4.
THe strange thing is, that with 2.3 it is working as it should.
Come on and solve this and let us know a patch for it.

**Comment by georgetze** [comment]
I just upgrade from magento 2.4.2  to 2.4.3  and it doesn't work for simple products either

**Comment by vseager** [comment]
@magento-engcom-team  This issue is still reproducible on 2.4.3 for both configurable or simple products.

**Comment by lytesaber** [comment]
As @vseager mentioned above this issue is still present on 2.4.3.

Admin Panel (Simple Product - Salable Quantity 0)
<img width="802" alt="image" src="https://user-images.githubusercontent.com/16328255/155262358-5249abe9-5efa-4b68-85dd-d07855cc6005.png">

Frontend (Simple Product - Stock Status "In Stock")
<img width="566" alt="image" src="https://user-images.githubusercontent.com/16328255/155262401-ec3f9e00-5cfa-4f45-90eb-148a7f6952f8.png">

Why has this issue been marked as resolved? And to add extra salt in to the wounds its listed as a fix in the Magento 2.4.3 release notes, what a joke!

<img width="539" alt="image" src="https://user-images.githubusercontent.com/16328255/155262676-6198fc06-e989-448a-874e-28f473864136.png">

It seems Adobe are more concerned with their issue close rate than issue resolutions and actual bugfixes.

Out of all the things in a ecommerce package how about we get the inventory system rock solid?

**Comment by tschirmer** [comment]
Still getting this on 2.4.3-p1

**Comment by MJTheOne** [comment]
Why is this closed? Still an issue

**Comment by gulshankumar** [comment]
Why this is closed?

**Comment by chequille** [comment]
I ask myself the same !!!!
Maybe nobody cares?

**Comment by sufyanrightpoint** [comment]
Reproducible in 2.4.4

**Comment by ugolek-work** [comment]
Reproduced in 2.4.5 when Default Stock is negative

**Comment by o0mohd0o** [comment]
Reproduced in 2.4.6 assigned source product qty = 0 . It shows in frontend that the product is in stock.

**Comment by stefanskotte** [comment]
> Reproduced in 2.4.5 when Default Stock is negative

Same here ! 

I have -75 reserved for a given color/size and it still shows as in stock on frontend. We do not allow for sales under 0 btw, so its not because of that.

This is pretty serious bug if you ask me.

**Comment by stefanskotte** [comment]
Did some further analysis on this. The field 'is_salable' comes from cataloginventory_stock_status.stock_status as part of the view creation for inventory_stock_x. During the create view this is considered as legacy stock, so I believe the actual support for reservations is implemented elsewhere.

I will look further into this.

**Comment by dphilipps** [comment]
Make sure all relevant consumers are running, in particular:

inventory.reservations.updateSalabilityStatus

**Comment by stefanskotte** [comment]
> Make sure all relevant consumers are running, in particular:
> 
> inventory.reservations.updateSalabilityStatus

Good point, let me check the current status of consumers.

**Comment by stefanskotte** [comment]
They look to be running fine, but as with consumers its a bit hard to figure that out fully:

<img width="1732" alt="image" src="https://user-images.githubusercontent.com/521625/235110901-604eff92-e404-46a9-ad5d-86648da2260b.png">

**Comment by stefanskotte** [comment]
@dphilipps I don't think the reservations table is the main problem, mine does the counts correctly - its more the aggregation and logic that ends up being determined in the isSalable() method that seems to be the problem.

**Comment by chequille** [comment]
Hi Guys,
there are a few other issues regarding this problems and what I said at the very beginning the problem is the method isSalable() and as well the other one is Saleable()

Both methods do check stock value and not salable quantity!
This is the major problem!

It does not have to do with consomers running or not running or all the other stuff mentioned in all the issue.

Purely isSalable() returns true if stock is > 0, but it should return true if salable quantity is > 0

This is the problem and nothing else as far as I see it.

Take care
Chequille

**Comment by stefanskotte** [comment]
> Hi Guys, there are a few other issues regarding this problems and what I said at the very beginning the problem is the method isSalable() and as well the other one isSaleable()
> 
> Both methods do check stock value and not salable quantity! This is the major problem!
> 
> It does not have to do with consomers running or not running or all the other stuff mentioned in all the issue.
> 
> Purely isSalable() returns true if stock is > 0, but it should return true if salable quantity is > 0
> 
> This is the problem and nothing else as far as I see it.
> 
> Take care Chequille

This is pretty much where I landed as well - I've been around the code, but it's abstracted to a level that makes no sense.

Getting an overview of what is actually going on and patching it properly isn't really straight forward.

Did you manage to find clues yourself in the code that could help me track down the negative reservations?

**Comment by chequille** [comment]
Hi Stefan,
I did two things:
1. I made the change as of this commit [372ccaa](https://github.com/magento/magento2/commit/372ccaa65c1bdc7b0ea814ffda2e0973cfd343c3) in app/code/Magento/InventoryIndexer/Model/IsProductSalable.php
2. Made my product page not cachable, as it only works if the page is not chached.

That is all I made and it is OK for me so far, but still not the solution, as this commit was never merged into any branch !?!?!?!?!

Regards,
Chequille

**Comment by stefanskotte** [comment]
I tried to implement the patch, but something changed on my 2.4.5-p1 on line 47

`        try {
            **$stockItem = $this->getStockItemData->execute($sku, $stockId);**
            $isSalable = (bool)($stockItem[GetStockItemDataInterface::IS_SALABLE] ?? false);
        } catch (LocalizedException $exception) {
            $this->logger->warning(
                sprintf(
                    'Unable to fetch stock #%s data for SKU %s. Reason: %s',
                    $stockId,
                    $sku,
                    $exception->getMessage()
                )
            );
            $isSalable = false;
        }`

**Comment by FadedOut** [comment]
To all still having an issue, I'm not sure if this information will help or not...but, I'd like to point to this potential solution: https://github.com/magento/magento2/issues/35319#issuecomment-1111842395

I'm using 2.4.6 but the bug I reported in that ticket still seems to be happening - so I did that fix which worked for me on 2.4.4, 2.4.5 and now on 2.4.6 and my swatches correctly represent out of stock products, so do simple products - they'll say out of stock if the backend shows 0 **SALABLE** qty. 

So maybe, hopefully, for the sake of all you poor souls still having this ridiculous issue, just doing that simple (one) file fix will solve your issue for you. Again, I'm using 2.4.6 (which that fix did work for 2.4.4 and 2.4.5 as well so any of those versions - not sure about before that though) and using that fix works great for me.

And don't get me started on why this **still** even needs to be "hacked" and isn't just fixed properly in Magento to begin with (it's been years), I'll never know. Some things I love Magento for (over other ecommerce platforms), then others downright disgust me (this being one).

**Comment by stefanskotte** [comment]
> To all still having an issue, I'm not sure if this information will help or not...but, I'd like to point to this potential solution: [#35319 (comment)](https://github.com/magento/magento2/issues/35319#issuecomment-1111842395)
> 
> I'm using 2.4.6 but the bug I reported in that ticket still seems to be happening - so I did that fix which worked for me on 2.4.4, 2.4.5 and now on 2.4.6 and my swatches correctly represent out of stock products, so do simple products - they'll say out of stock if the backend shows 0 **SALABLE** qty.
> 
> So maybe, hopefully, for the sake of all you poor souls still having this ridiculous issue, just doing that simple (one) file fix will solve your issue for you. Again, I'm using 2.4.6 (which that fix did work for 2.4.4 and 2.4.5 as well so any of those versions - not sure about before that though) and using that fix works great for me.
> 
> And don't get me started on why this **still** even needs to be "hacked" and isn't just fixed properly in Magento to begin with (it's been years), I'll never know. Some things I love Magento for (over other ecommerce platforms), then others downright disgust me (this being one).

I have implemented this patch now for our 2.4.5-p1 running size, color, length attributes on configurable products.

Will refresh now and see where it goes.

**Comment by stefanskotte** [comment]
@FadedOut 

It seems to work with salable quantity = 0, but it still won't see negative quantity (eg. -1) as out of stock, and thus displays the swatch for it.

I have set "Backorders: No Backorders" and out of stock threshold is at default 0.

So I'm guessing we're still back at Magento seeing anything but 0 as in stock......

**Comment by FadedOut** [comment]
@stefanskotte 

In that case, I would open a new ticket with that information (and explaining that you had to do that hack I posted because that should have been fixed in 2.4.6 but it seems it wasn't). 

But, can I ask what the purpose of having a negative stock would be if you're not doing backorders? If you can, just simply set them to 0 instead?
